atomic_cell.cc

```diff
@@ -238,8 +241,21 @@ operator<<(std::ostream& os, const atomic_cell_view::printer& acvp) {
     auto& type = acvp._type;
     auto& acv = acvp._cell;
     if (acv.is_live()) {
+        std::ostringstream cell_value_string_builder;
+        if (type.is_counter()) {
+            if (acv.is_counter_update()) {
+                cell_value_string_builder << "counter_update_value=" << acv.counter_update_value();
+            } else {
+                cell_value_string_builder << "shards: ";
+                counter_cell_view::with_linearized(acv, [&cell_value_string_builder] (counter_cell_view& ccv) {
+                    cell_value_string_builder << ::join(", ", ccv.shards());
+                });
+            }
+        } else {
+            cell_value_string_builder << type.to_string(acv.value().linearize());
+        }
         return fmt_print(os, "atomic_cell{{{},ts={:d},expiry={:d},ttl={:d}}}",
-            type.to_string(acv.value().linearize()),
+            cell_value_string_builder.str(),
```

Comment:
Why not fix to_string() for counter types?

Right now it calls fail(unimlemented).

Reply:
I was thinking about it, but found no easy way. BTW. abort comes from:
https://github.com/scylladb/scylla/blob/a16cb0071946f6a08d77e26c7eaff092ea375a4e/data/cell.hh#L651

Cell here goes as `bytes`, through `deserialize_visitor` and `to_string_impl_visitor`. In serialized representation we don't know whether it's a "delta" or "shards". Maybe it can be worked around (e.g. by guessing from buffer size), but this idea was worrying me.

Moreover, I think `counter_type_impl` is not prepared to store "shards". At least I don't see at this point how to squeeze "shards" and "deltas" into one concrete_type. I'll accept any advice here.

And... it was tempting with simplicity - without hijacking `imr::pod<int64_t>` in `atomic_cell_view::value()`, without serializing back and forth...

Reply:
No, my idea is completely wrong. counter_type_impl is the front-end's view of a counter, which is just an int64_t. atomic_cell uses the back-end's view of a counter, which has all these shard things. Usually the front-end and the back-end share the same view, but for counters, they don't.